src/util/cmdline.cpp

```diff
@@ -20,175 +21,333 @@ cmdlinet::~cmdlinet()
 
 void cmdlinet::clear()
 {
-  options.clear();
+  vm.clear();
   args.clear();
 }
 
-bool cmdlinet::isset(char option) const
-{
-  int i;
-
-  i = getoptnr(option);
-  if(i < 0)
-    return false;
-  return options[i].isset;
-}
-
 bool cmdlinet::isset(const char *option) const
 {
-  int i;
-
-  i = getoptnr(option);
-  if(i < 0)
-    return false;
-  return options[i].isset;
+  return vm.count(option);
 }
 
-const char *cmdlinet::getval(char option) const
-{
-  int i;
-
-  i = getoptnr(option);
-  if(i < 0)
-    return (const char *)nullptr;
-  if(options[i].values.empty())
-    return (const char *)nullptr;
-  return options[i].values.front().c_str();
-}
-
-const std::list<std::string> &cmdlinet::get_values(char option) const
+const std::list<std::string> &cmdlinet::get_values(const char *option) const
 {
-  int i;
-
-  i = getoptnr(option);
-  assert(i >= 0);
-  return options[i].values;
+  auto &value = vm[option].value();
+  if(auto v = boost::any_cast<int>(&value))
+  {
+    auto res = new std::list<std::string>();
+    res->emplace_front(std::to_string(*v));
+    return *res;
+  }
+  if(auto v = boost::any_cast<std::string>(&value))
+  {
+    auto res = new std::list<std::string>();
+    res->emplace_front(*v);
+    return *res;
+  }
+  auto src = vm[option].as<std::vector<std::string>>();
+  auto dest = new std::list<std::string>(src.begin(), src.end());
+  return *dest;
 }
 
 const char *cmdlinet::getval(const char *option) const
 {
-  int i;
-
-  i = getoptnr(option);
-  if(i < 0)
-    return (const char *)nullptr;
-  if(options[i].values.empty())
+  if(!vm.count(option))
     return (const char *)nullptr;
-  return options[i].values.front().c_str();
-}
-
-const std::list<std::string> &cmdlinet::get_values(const char *option) const
-{
-  int i;
-
-  i = getoptnr(option);
-  assert(i >= 0);
-  return options[i].values;
-}
-
-int cmdlinet::getoptnr(char option) const
-{
-  for(unsigned i = 0; i < options.size(); i++)
-    if(options[i].optchar == option)
-      return i;
-
-  return -1;
-}
-
-int cmdlinet::getoptnr(const char *option) const
-{
-  for(unsigned i = 0; i < options.size(); i++)
-    if(options[i].optstring == option)
-      return i;
-
-  return -1;
+  auto &value = vm[option].value();
+  if(auto v = boost::any_cast<int>(&value))
+  {
+    auto res = new std::string(std::to_string(*v));
+    return res->c_str();
+  }
+  if(auto v = boost::any_cast<std::string>(&value))
+    return v->c_str();
+  if(auto v = boost::any_cast<std::vector<std::string>>(&value))
+    return v->front().c_str();
+  return (const char *)nullptr;
 }
 
-bool cmdlinet::parse(int argc, const char **argv, const struct opt_templ *opts)
```

Comment:
Please keep the boolean return value.

Reply:
The function boolean output no longer signifies errors of command line inputs, instead an exception is thrown from store function call within parse if any error occurs. These errors include unrecognized options, invalid values, option argument required value as well as other validations.
Hence, a try catch around parse call will be sufficient to take care of capturing the error message and presenting it to the user.

Reply:
I see, so we can simplify the patch even further:
1. The `try/catch` is only there to handle exceptions thrown by the `store` method? If so, add the `try/catch` only to cover the `store` method call, return `false` if there was no issue, print the exception.what() to `cerr`, and return `true` otherwise.
2. Bring back the boolean return and the class member
3. In `parseoptions_baset::main`, only return `EX_USAGE` if the boolean flag is true (printing the message was done in 1.)

That way we can keep the old workflow of returning booleans and we don't have to store the error message.

BTW, shouldn't the `store` be scoped, i.e., `boost::something::store`?